ql/pom.xml

```diff
@@ -880,6 +880,12 @@
                   <include>joda-time:joda-time</include>
                   <include>org.apache.calcite:*</include>
                   <include>org.apache.calcite.avatica:avatica</include>
+                  <!-- include all of calcite's transitive dependencies as well since
+                    otherwise they'll be removed in assembly packaging -->
+                  <include>net.hydromatic:eigenbase-properties</include>
+                  <include>org.codehaus.janino:janino</include>
+                  <include>org.codehaus.janino:commons-compiler</include>
+                  <include>org.pentaho:pentaho-aggdesigner-algorithm</include>
```

Comment:
One question. Will this be flaky? Once calcite changes transitive dependencies, seems we also need to update this.

Reply:
Yeah that's the downside of it - whenever we upgrade calcite we should check its dependencies and update here as well. I'm not sure if there is an easy way to include an artifact and all its transitive dependencies in the maven-shade-plugin

Reply:
BTW, I remember in the latest branch, the shading guava PR doesn't include transitive dependencies like that. Do we also need to do it for the branch?

Reply:
I still need to verify whether there is issue in master/4.0.0 as well - it uses calcite 1.21 which no longer depend on eigenbase which causes the issue in our case, but I think it might be good to always include transitive dependencies anyway.

Reply:
So I tried master branch and didn't hit the issue. However I think I may still open another PR for it since we don't really when the transitive dependencies will be needed at runtime.

Reply:
Oh actually in master we already handles this [here](https://github.com/apache/hive/blob/master/pom.xml#L1052), as part of [HIVE-22126](https://issues.apache.org/jira/browse/HIVE-22126). I think we should follow the same approach and explicitly add the transitive dependency in Hive's top-level pom. I think this is also how Spark manages its transitive dependencies? it might be better than pulling out the classes and putting them in the fat jar.